ledger/complete/mtrie/trieCache.go

```diff
@@ -0,0 +1,149 @@
+package mtrie
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+type OnTreeEvictedFunc func(tree *trie.MTrie)
+
+// TrieCache caches tries into memory, it acts as a fifo queue
+// so when it reaches to the capacity it would evict the oldest trie
+// from the cache.
+//
+// Under the hood it uses a circular buffer
+// of mtrie pointers and a map of rootHash to cache index for fast lookup
+type TrieCache struct {
+	tries         []*trie.MTrie
+	lookup        map[ledger.RootHash]int // index to item
+	lock          sync.RWMutex
+	capacity      int
+	tail          int // element index to write to
+	count         int // number of elements (count <= capacity)
+	onTreeEvicted OnTreeEvictedFunc
+}
+
+// NewTrieCache returns a new TrieCache with given capacity.
+func NewTrieCache(capacity uint, onTreeEvicted OnTreeEvictedFunc) *TrieCache {
+	return &TrieCache{
+		tries:         make([]*trie.MTrie, capacity),
+		lookup:        make(map[ledger.RootHash]int, capacity),
+		lock:          sync.RWMutex{},
+		capacity:      int(capacity),
+		tail:          0,
+		count:         0,
+		onTreeEvicted: onTreeEvicted,
+	}
+}
+
+// Purge removes all mtries stored in the buffer
+func (tc *TrieCache) Purge() {
+	tc.lock.Lock()
+	defer tc.lock.Unlock()
+
+	tc.tail = 0
+	tc.count = 0
+	for i := 0; i < tc.capacity; i++ {
+		if tc.onTreeEvicted != nil {
+			if tc.tries[i] != nil {
+				tc.onTreeEvicted(tc.tries[i])
+			}
+		}
+		tc.tries[i] = nil
+	}
+	tc.lookup = make(map[ledger.RootHash]int, tc.capacity)
+}
+
+// Tries returns elements in queue, starting from the oldest element
+// to the newest element.
+func (tc *TrieCache) Tries() []*trie.MTrie {
+	tc.lock.RLock()
+	defer tc.lock.RUnlock()
+
+	if tc.count == 0 {
+		return nil
+	}
+
+	tries := make([]*trie.MTrie, tc.count)
+
+	if tc.isFull() {
```

Comment:
just a random though: if we used a double-linked list, would the logic here be a bit simpler?

Reply:
array is much link list

Reply:
agree!  that said, given that we use map for O(1) access it should not affect perf too much (esp. given how infrequently we should update the forest.)

Reply:
there is going to be a follow up PR for improvements, going to consider your suggestions in the PR.